### PR TITLE
MultiCluster: rework on nodePort type gateway service

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
This is a follow-up optimization after #22885, and also fix a legacy bug when getting gateway address.

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 1]
```diff
@@ -467,24 +463,48 @@ func (c *Controller) Stop() {
...
-		}
-	}
-
 	c.Lock()
 	out := make([]*model.Service, 0, len(c.servicesMap))
 	for _, svc := range c.servicesMap {
-		if isNodePortService(svc) {
+		out = append(out, svc)
+	}
+	c.Unlock()
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:471]: Please change this back to RUnlock

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 2]
```diff
@@ -467,24 +463,48 @@ func (c *Controller) Stop() {
...
+			for _, i := range rawNodes {
+				machine := i.(*v1.Node)
+				for _, address := range machine.Status.Addresses {
+					if address.Type == v1.NodeExternalIP && address.Address != "" {
+						k8sNodes = append(k8sNodes,
+							&kubernetesNode{address: address.Address, labels: machine.Labels})
+						extAddresses = append(extAddresses, address.Address)
+					}
+				}
+			}
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:506]: This logic needs to be moved out. It’s one time computation for all services

File: pilot/pkg/serviceregistry/kube/conversion.go [hunk 3]
```diff
@@ -115,30 +119,35 @@ func ConvertService(svc coreV1.Service, domainSuffix string, clusterID string) *
 		},
 	}
 
-	if svc.Spec.Type == coreV1.ServiceTypeNodePort {
+	switch svc.Spec.Type {
+	case coreV1.ServiceTypeNodePort:
+		if svc.Annotations[NodeSelectorAnnotation] == "" {
+			// only do this for istio ingress-gateway services
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/kube/conversion.go:125]: How can the user specify pick all nodes?

File: pilot/pkg/serviceregistry/aggregate/controller.go [hunk 4]
```diff
@@ -144,19 +144,6 @@ func (c *Controller) Services() ([]*model.Service, error) {
...
-						sp.Attributes.ClusterExternalAddresses = make(map[string][]string)
-					}
-					sp.Attributes.ClusterExternalAddresses[r.Cluster()] = s.Attributes.ClusterExternalAddresses[r.Cluster()]
-				}
-				if s.Attributes.ClusterExternalPorts != nil && len(s.Attributes.ClusterExternalPorts[r.Cluster()]) > 0 {
-					if sp.Attributes.ClusterExternalPorts == nil {
-						sp.Attributes.ClusterExternalPorts = make(map[string]map[uint32]uint32)
-					}
-					sp.Attributes.ClusterExternalPorts[r.Cluster()] = s.Attributes.ClusterExternalPorts[r.Cluster()]
-				}
```
Comment (rshriram) [inline pilot/pkg/serviceregistry/aggregate/controller.go:159]: Why remove this? the data for a service returned by both services() and getService() should be consistent. Besides this is not heavy computation right? It’s just a cheap map merging.

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/aggregate/controller.go:159]: This is a legacy bug, we only GetService when calc gateway address

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:506]: I donot know what i was thinking

File: pilot/pkg/serviceregistry/kube/conversion.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/conversion.go:125]: got it, though it is not recommended to do use it like this.

File: pilot/pkg/serviceregistry/aggregate/controller.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/aggregate/controller.go:159]: The external port and address only used for gateway address cal

Comment (rshriram) [reply pilot/pkg/serviceregistry/aggregate/controller.go:159]: That is true.. but for future use, any caller expecting to see the cluster external address in the service object will be confused as to why it is not appearing. This is not a costly merge right? So does it matter if we have this logic here or not?

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/aggregate/controller.go:159]: I meant to merge Services() and GetService() once i am free

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/aggregate/controller.go:159]: But if you insist on, I can restore it

Comment (rshriram) [reply pilot/pkg/serviceregistry/aggregate/controller.go:159]: never mind about this code for now.

**Pull request merged**